### PR TITLE
✨ use local_mode with cspo

### PR DESCRIPTION
**Pull request by kranurag7**
This PR adds local_mode with cspo. 
This also removes helm from cspo image. (Fixes https://github.com/SovereignCloudStack/cluster-stack-provider-openstack/issues/81)

There are some follow-up task (usage docs) to this which can be added in a separate commit or PR.

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -17,20 +17,21 @@ settings = {
...
     "cert_manager_version": "v1.13.1",
     "kustomize_substitutions": {
     },
 }
 
 # global settings
-settings.update(read_json(
-    "tilt-settings.json",
+settings.update(read_yaml(
+    "tilt-settings.yaml",
```
Comment (chess-knight) [inline Tiltfile:34]: Do not forget .gitignore please

File: Tiltfile [hunk 2]
```diff
@@ -124,14 +181,19 @@ def fixup_yaml_empty_arrays(yaml_str):
     return yaml_str.replace("storedVersions: null", "storedVersions: []")
 
 ## This should have the same versions as the Dockerfile
-tilt_dockerfile_header_cspo = """
-FROM docker.io/alpine/helm:3.12.2 as helm
```
Comment (chess-knight) [inline Tiltfile:128]: Thanks for removing Helm, but I suggest removing Helm from the whole repository(e.g. Makefile, README.md, .gitignore, .github/)

File: Tiltfile [hunk 3]
```diff
@@ -166,18 +228,31 @@ def deploy_cspo():
...
+            target = "tilt",
+            entrypoint = entrypoint,
+            live_update = [
+                sync(".tiltbuild/manager", "/manager"),
+            ],
+            ignore = ["templates"],
+        )
     k8s_yaml(blob(yaml))
     k8s_resource(workload = "cspo-controller-manager", labels = ["cspo"])
     k8s_resource(
```
Comment (chess-knight) [inline Tiltfile:258]: I suggest to add to objects here all the uncategorized objects, e.g. `openstackclusterstackreleases.infrastructure.clusterstack.x-k8s.io:customresourcedefinition`...

File: Tiltfile [hunk 4]
```diff
@@ -61,12 +62,68 @@ def deploy_capi():
             if kb_extra_args:
                 patch_args_with_extra_args("capi-kubeadm-bootstrap-system", "capi-kubeadm-bootstrap-controller-manager", kb_extra_args)
 
+tilt_dockerfile_header_cso = """
+FROM ghcr.io/sovereigncloudstack/cso:v0.1.0-alpha.3 as builder
```
Comment (chess-knight) [inline Tiltfile:66]: Can the settings.get("cso_version") be used here?

File: Tiltfile [hunk 5]
```diff
@@ -61,12 +62,68 @@ def deploy_capi():
...
+    # Set up an image build for the provider. The live update configuration syncs the output from the local_resource
+    # build into the container.
+    docker_build_with_restart(
+        ref = "ghcr.io/sovereigncloudstack/cso-test",
+        context = ".",
+        dockerfile_contents = tilt_dockerfile_header_cso,
+        target = "tilt",
+        entrypoint = entrypoint,
+        live_update = [
+            sync("./local_cso.yaml", "/local_cso.yaml"), # reload when we change the manifest
```
Comment (chess-knight) [inline Tiltfile:103]: I suggest adding this also to the `tilt_dockerfile_header_cso`

File: Tiltfile [hunk 6]
```diff
@@ -208,8 +283,7 @@ def cspo_template():
     cspo_yaml = local(cmd, quiet=True)
     k8s_yaml(cspo_yaml)
     k8s_resource(objects = ["cspotemplate:openstackclusterstackreleasetemplate"], new_name = "cspotemplate", labels = ["cspo-template"])
-
-
+    
 def clusterstack():
     k8s_resource(objects = ["clusterstack:clusterstack"], new_name = "clusterstack", labels = ["clusterstack"])
```
Comment (chess-knight) [inline Tiltfile:288]: I don't know why, but in the local mode this clusterstack is red and needs retrigger

**Review by chess-knight — COMMENTED** [review]
Thanks for this PR with updated CSO, I left multiple comments/questions.
Can you add also docs, please?

File: Tiltfile (hunk 6 above)
Comment (kranurag7) [reply Tiltfile:288]: Yes, I was not able to find a good solution for it. It's happening because we are applying clusterstack too early before the CRDs are applied.

Comment (chess-knight) [reply Tiltfile:288]: Did you try the same way as cspo_template() is done?

Comment (kranurag7) [reply Tiltfile:288]: Roman, I didn't gave much attention to this one. 
I created an issue here https://github.com/SovereignCloudStack/cluster-stack-provider-openstack/issues/91 

Let's do it in a separate PR, this can be handled separately. This is not a blocker for me as of now in using tilt setup. I'll do a separate PR for this.

Comment (chess-knight) [reply Tiltfile:288]: Thank you!

File: Tiltfile (hunk 4 above)
Comment (kranurag7) [reply Tiltfile:66]: updated the same like this and it's working. 

```
tilt_dockerfile_header_cso = """
FROM ghcr.io/sovereigncloudstack/cso:{} as builder

FROM docker.io/library/alpine:3.18.0 as tilt
WORKDIR /
COPY --from=builder /usr/local/bin/helm /usr/local/bin/helm
COPY --from=builder /manager /manager
COPY .release/ /tmp/downloads/cluster-stacks/
""".format(settings.get("cso_version"))
```

File: local_cso.yaml [hunk 7]
```diff
@@ -0,0 +1,940 @@
...
+        cluster.x-k8s.io/provider: infrastructure-cluster-stack-operator
+        control-plane: cso-controller-manager
+    spec:
+      containers:
+      - args:
+        - --leader-elect=true
+        - --local=true
+        command:
+        - /manager
+        image: ghcr.io/sovereigncloudstack/cso-test:v0.1.0-alpha.3
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply local_cso.yaml:874]: Thank you for the explanation!

File: Tiltfile [hunk 8]
```diff
@@ -61,12 +62,68 @@ def deploy_capi():
...
+
 def deploy_cso():
     version = settings.get("cso_version")
     cso_uri = "https://github.com/sovereignCloudStack/cluster-stack-operator/releases/download/{}/cso-infrastructure-components.yaml".format(version)
     cmd = "curl -sSL {} | {} | kubectl apply -f -".format(cso_uri, envsubst_cmd)
     local(cmd, quiet = True)
 
+def deploy_local_cso():
+    yaml_cso = './local_cso.yaml'
+    local_resource(
```
[the remark this answers is not in the document]
Comment (kranurag7) [reply Tiltfile:83]: We were using local_resource to use `sed` and `envsubst` operation but given we are not using it anymore, I'm removing local_resource from `deploy_local_cso()` function. 

`live_update` rebuilds/reloads the content the inside the container from host system when we change anything in things we are targeting using `sync`. 

From more details on how does it work: 
Ref: https://github.com/tilt-dev/tilt-extensions/tree/master/restart_process#whats-happening-under-the-hood 
TLDR is it uses `entr` for looking at changes in the underlying files using inotify and builds/loads it again. AFAIK.

File: Tiltfile (hunk 5 above)
Comment (kranurag7) [reply Tiltfile:103]: updated `COPY local_cso.yaml /local_cso.yaml`

File: Tiltfile (hunk 2 above)
Comment (chess-knight) [reply Tiltfile:128]: Please remove from .gitignore also

File: Tiltfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (chess-knight) [reply Tiltfile]: Can you also update the comment above `docker_build_with_restart` that I mentioned in this conversation?

File: Tiltfile (hunk 6 above)
Comment (chess-knight) [reply Tiltfile:284]: During testing, I see cspotemplate again red and needs to be triggered, can you add that to issue 91 as well, please?

Comment (chess-knight) [reply Tiltfile:284]: Maybe after adding `openstackclusterstackreleasetemplates` to k8s_resource objects it happened? I don't know

File: Tiltfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (kranurag7) [reply Tiltfile]: Sure, I'm removing the comment in the next commit.

File: Tiltfile (hunk 6 above)
Comment (kranurag7) [reply Tiltfile:284]: Yes, I noticed it happened again but we are going to apply the templates in the subsequent step after the controller starts. 
Tilt will only start the controller and then we apply the resources using a click that will be backed by a make target.

**Review by chess-knight — APPROVED** [review]
LGTM

**Review by michal-gubricky — APPROVED** [review]
